File: tests/src/joda-time/joda-time/2.9.9/src/test/java/joda_time/joda_time/DateTimeZoneTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package joda_time.joda_time;

import static org.assertj.core.api.Assertions.assertThat;

import org.joda.time.DateTimeZone;
import org.joda.time.tz.DefaultNameProvider;
import org.joda.time.tz.NameProvider;
import org.joda.time.tz.Provider;
import org.joda.time.tz.UTCProvider;
import org.junit.jupiter.api.Test;

public class DateTimeZoneTest {

    @Test
    void loadsProviderFromConfiguredSystemProperty() {
        String propertyName = "org.joda.time.DateTimeZone.Provider";
        String previousPropertyValue = System.getProperty(propertyName);
        Provider previousProvider = DateTimeZone.getProvider();

        try {
            System.setProperty(propertyName, UTCProvider.class.getName());

            DateTimeZone.setProvider(null);

            assertThat(DateTimeZone.getProvider()).isInstanceOf(UTCProvider.class);
            assertThat(DateTimeZone.getProvider().getAvailableIDs()).containsExactly("UTC");
        } finally {
            restoreProperty(propertyName, previousPropertyValue);
            DateTimeZone.setProvider(previousProvider);
        }
    }

    @Test
    void loadsNameProviderFromConfiguredSystemProperty() {
        String propertyName = "org.joda.time.DateTimeZone.NameProvider";
        String previousPropertyValue = System.getProperty(propertyName);
        NameProvider previousNameProvider = DateTimeZone.getNameProvider();

        try {
            System.setProperty(propertyName, DefaultNameProvider.class.getName());

            DateTimeZone.setNameProvider(null);

            assertThat(DateTimeZone.getNameProvider()).isInstanceOf(DefaultNameProvider.class);
        } finally {
            restoreProperty(propertyName, previousPropertyValue);
            DateTimeZone.setNameProvider(previousNameProvider);
        }
    }

    private static void restoreProperty(String propertyName, String propertyValue) {
        if (propertyValue == null) {
            System.clearProperty(propertyName);
        } else {
            System.setProperty(propertyName, propertyValue);
        }
    }
}

File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/DateTimeZoneTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import java.util.Locale
import java.util.Set

import org.joda.time.DateTimeZone
import org.joda.time.tz.NameProvider
import org.joda.time.tz.Provider
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test

class DateTimeZoneTest {
  @Test
  def systemPropertiesConfigureProviderAndNameProvider(): Unit = {
    val originalProvider: Provider = DateTimeZone.getProvider
    val originalNameProvider: NameProvider = DateTimeZone.getNameProvider
    val originalProviderProperty: String = System.getProperty(providerPropertyName)
    val originalNameProviderProperty: String = System.getProperty(nameProviderPropertyName)

    try {
      System.setProperty(providerPropertyName, classOf[ConfiguredDateTimeZoneProvider].getName)
      System.setProperty(nameProviderPropertyName, classOf[ConfiguredDateTimeZoneNameProvider].getName)

      DateTimeZone.setProvider(null)
      DateTimeZone.setNameProvider(null)

      val provider: Provider = DateTimeZone.getProvider
      val nameProvider: NameProvider = DateTimeZone.getNameProvider

      assertEquals(classOf[ConfiguredDateTimeZoneProvider], provider.getClass)
      assertEquals(classOf[ConfiguredDateTimeZoneNameProvider], nameProvider.getClass)
      assertSame(DateTimeZone.UTC, provider.getZone("UTC"))
      assertEquals("UTC", provider.getAvailableIDs.iterator().next())
      assertEquals("TST", DateTimeZone.UTC.getShortName(0L, Locale.ENGLISH))
      assertEquals("Test time", DateTimeZone.UTC.getName(0L, Locale.ENGLISH))
    } finally {
      restoreProperty(providerPropertyName, originalProviderProperty)
      restoreProperty(nameProviderPropertyName, originalNameProviderProperty)
      DateTimeZone.setProvider(originalProvider)
      DateTimeZone.setNameProvider(originalNameProvider)
    }
  }

  private def providerPropertyName: String = "org.joda.time.DateTimeZone.Provider"

  private def nameProviderPropertyName: String = "org.joda.time.DateTimeZone.NameProvider"

  private def restoreProperty(name: String, value: String): Unit = {
    if (value == null) {
      System.clearProperty(name)
    } else {
      System.setProperty(name, value)
    }
  }
}

final class ConfiguredDateTimeZoneProvider extends Provider {
  override def getZone(id: String): DateTimeZone = {
    if ("UTC" == id) {
      DateTimeZone.UTC
    } else {
      null
    }
  }

  override def getAvailableIDs: Set[String] = Set.of("UTC")
}

final class ConfiguredDateTimeZoneNameProvider extends NameProvider {
  override def getShortName(locale: Locale, id: String, nameKey: String): String = "TST"

  override def getName(locale: Locale, id: String, nameKey: String): String = "Test time"
}
